src/utils/daos_hdlr.c

```diff
@@ -2566,95 +2576,266 @@ fs_copy(struct file_dfs *src_file_dfs,
+		dpe_types[1] = ca->cont_prop_oid;
+
+		rc = dm_get_cont_prop(ca->src_coh, sysname, src_cont_info, size,
+				      dpe_types, dpe_vals);
 		if (rc != 0) {
-			fprintf(stderr, "failed to connect to destination "
-				"pool: %d\n", rc);
+			fprintf(stderr, "failed to query source "
+				"container: %d\n", rc);
+			D_GOTO(err_src, rc);
+		}
+
+		ca->cont_layout = dpe_vals[0];
+		ca->cont_oid = dpe_vals[1];
+
+		props = daos_prop_alloc(2);
+		if (props == NULL) {
+			fprintf(stderr, "Failed to allocate prop (%d)", rc);
 			D_GOTO(out, rc);
 		}
-		rc = daos_cont_open(fa->dst_poh, fa->dst_c_uuid, DAOS_COO_RW,
-				    &fa->dst_coh, dst_cont_info, NULL);
-		if (rc != 0) {
-			fprintf(stderr, "failed to open destination "
-				"container: %d\n", rc);
-			D_GOTO(err_dst_root, rc);
+		props->dpp_entries[0].dpe_type = ca->cont_prop_layout;
+		props->dpp_entries[0].dpe_val = ca->cont_layout;
+
+		/* only set max oid on created dest containers
+		 * if this is a non-posix copy
+		 */
+		if (!is_posix_copy) {
+			props->dpp_entries[1].dpe_type = ca->cont_prop_oid;
+			props->dpp_entries[1].dpe_val = ca->cont_oid;
 		}
-		rc = dfs_mount(fa->dst_poh, fa->dst_coh, O_RDWR,
-			       &dst_file_dfs->dfs);
-		if (rc != 0) {
-			fprintf(stderr, "dfs mount on destination "
-			"failed: %d\n", rc);
-			D_GOTO(err_dst_dfs, rc);
+	}
+
+	/* open dst pool, dst cont, and mount dfs */
+	if (dst_file_dfs->type == DAOS) {
+		if (!uuid_is_null(ca->dst_p_uuid)) {
+			rc = daos_pool_connect(ca->dst_p_uuid, sysname,
+					       DAOS_PC_RW, &ca->dst_poh, NULL,
+					       NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to connect to "
+					"destination pool: %d\n", rc);
+				D_GOTO(out, rc);
+			}
+		/* if the dst pool uuid is null that means that this
+		 * is a UNS destination path, so we copy the source
+		 * pool uuid into the destination and try to connect
+		 * again
+		 */
+		} else {
+			uuid_copy(ca->dst_p_uuid, ca->src_p_uuid);
+			rc = daos_pool_connect(ca->dst_p_uuid, sysname,
+					       DAOS_PC_RW, &ca->dst_poh,
+					       NULL, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to connect to "
+					"destination pool: %d\n", rc);
+				D_GOTO(err_src, rc);
+			}
+			uuid_copy(dattr.da_puuid, ca->dst_p_uuid);
+			uuid_copy(dattr.da_cuuid, ca->dst_c_uuid);
+			dattr.da_type = ca->cont_layout;
+			dattr.da_props = props;
+			rc = duns_create_path(ca->dst_poh, path,
+					      &dattr);
+			if (rc != 0) {
+				fprintf(stderr, "provide a destination "
+					"pool or UNS path of the form:\n\t"
+					"--dst </$pool> | </path/to/uns>\n");
+					D_GOTO(err_dst_root, rc);
+			}
+			uuid_copy(ca->dst_c_uuid, dattr.da_cuuid);
+		}
+		/* try to open container if this is a filesystem copy,
+		 * and if it fails try to create a destination,
+		 * then attempt to open again
+		 */
+		rc = daos_cont_open(ca->dst_poh, ca->dst_c_uuid,
+				    DAOS_COO_RW, &ca->dst_coh,
+				    dst_cont_info, NULL);
+		if (rc == -DER_NONEXIST) {
```

Comment:
@dsikich I think we need an else for this, right?
Something like:
```
if (rc == -DER_NONEXIST) {
    // create it and try to open again
} else {
    // error
}
```

Reply:
@daltonbohning well that isn't really an error, since if it successfully opens then it isn't an error, but I suppose if rc != 0 at that point it is so I can check

Reply:
@dsikich Yes, sorry, that's what I meant to say. If rc is 0 and it's also not -DER_NONEXIST, then that is a real error.